### PR TITLE
WFLY-10784 not fail WeldStartCompletionService on optional components

### PR DESCRIPTION
**Pull request by kudrevatykh**
Thanks for submitting your Pull Request!

Please make sure your PR meets the following requirements:
- [x ] Pull Request title is properly formatted: `[WFLY-XYZ] Subject` or `WFLY-XYZ Subject`
- [ x] Pull Request contains link to the JIRA issue(s)
- [ x] Pull Request contains description of the issue(s)
- [ x] Pull Request does not include fixes for issues other than the main ticket
- [ x] Attached commits represent units of work and are properly formatted

For bigger changes, major and minor component upgrades make sure your PR also meets following requirements:
- [ ] Pull Request requires a change to the documentation
- [ ] Documentation have been updated accordingly
- [ ] Tests were added to cover changes

For new features ensure as well:
- [ ] Analysis was done
- [ ] Test Plan has been done
- [ ] Tests were verified in advance

If you are not an active contributor of the WildFly project you can request sponsorship by one of the members to help guide you through the process.

### CODE REVIEW

File: weld/subsystem/src/main/java/org/jboss/as/weld/deployment/processors/WeldDeploymentProcessor.java [hunk 1]
```diff
@@ -293,18 +294,20 @@ public void deploy(DeploymentPhaseContext phaseContext) throws DeploymentUnitPro
         ServiceBuilder<WeldStartCompletionService> weldStartCompletionServiceBuilder = serviceTarget
                 .addService(weldStartCompletionServiceName, weldStartCompletionService)
                 .addDependency(weldBootstrapServiceName, WeldBootstrapService.class, weldStartCompletionService.getBootstrap())
-                .addDependency(weldStartServiceName).addDependencies(getComponentStartServiceNames(deploymentUnit));
+                .addDependency(weldStartServiceName);
+        getComponentStartServiceNames(deploymentUnit).entrySet().forEach(e->weldStartCompletionServiceBuilder.addDependencies(e.getKey(), e.getValue()));
         for (DeploymentUnit sub : subDeployments) {
-            weldStartCompletionServiceBuilder.addDependencies(getComponentStartServiceNames(sub));
+            getComponentStartServiceNames(sub).entrySet().forEach(e->weldStartCompletionServiceBuilder.addDependencies(e.getKey(), e.getValue()));
```
Comment (manovotn) [inline weld/subsystem/src/main/java/org/jboss/as/weld/deployment/processors/WeldDeploymentProcessor.java]: `ServiceBuilder.addDependencies()` is deprecated and even says that optional dependencies are unsafe and should not be used.

File: weld/subsystem/src/main/java/org/jboss/as/weld/deployment/processors/WeldDeploymentProcessor.java [hunk 2]
```diff
@@ -321,9 +324,9 @@ public void deploy(DeploymentPhaseContext phaseContext) throws DeploymentUnitPro
     private void installBootstrapConfigurationService(WeldDeployment deployment, DeploymentUnit parentDeploymentUnit) {
         final boolean nonPortableMode = parentDeploymentUnit.getAttachment(WeldConfiguration.ATTACHMENT_KEY).isNonPortableMode();
         final ExternalConfiguration configuration = new ExternalConfigurationBuilder()
-            .add(ConfigurationKey.NON_PORTABLE_MODE.get(), nonPortableMode)
-            .add(ConfigurationKey.ALLOW_OPTIMIZED_CLEANUP.get(), true)
-            .build();
+                .add(ConfigurationKey.NON_PORTABLE_MODE.get(), nonPortableMode)
```
Comment (manovotn) [inline weld/subsystem/src/main/java/org/jboss/as/weld/deployment/processors/WeldDeploymentProcessor.java]: Unnecessary formatting.

**Review by manovotn — CHANGES_REQUESTED** [review]
@kudrevatykh first and foremost, I don't understand what is the case where current solution fails (even after reading the [JIRA issue](https://issues.jboss.org/browse/WFLY-10784)). Can you provide a reproducer, detailed description, or, better still, add an automated test?

Secondly, the PR uses deprecated methods from `ServiceBuilder` which rely on `DependencyType` and according to javadoc - "_Optional dependencies are unsafe and should not be used._". So this isn't a way to go.

We first need to see what's the actual problem and have a way of reproducing it, then we can jump to solutions. That being said, we will likely need to update `WeldDeploymentProcessor` anyway to remove usage of deprecated classes.

**Comment by kabir** [comment]
@kudrevatykh Adding the fixme label until you have addressed @manovotn 's concerns. @mention me once it is all in order

File: weld/subsystem/src/main/java/org/jboss/as/weld/deployment/processors/WeldDeploymentProcessor.java (hunk 1 above)
Comment (kudrevatykh) [reply weld/subsystem/src/main/java/org/jboss/as/weld/deployment/processors/WeldDeploymentProcessor.java]: Can you provide recipe how to depend on component when it can not start in any moment?

Comment (manovotn) [reply weld/subsystem/src/main/java/org/jboss/as/weld/deployment/processors/WeldDeploymentProcessor.java]: I don't know if it is possible at all. With current setup, Weld can do some cleanup operation allowing us to lower the footprint. If we cannot rely on all components started, then we cannot do any cleanup. It also seems that `org.jboss.msc.service.ServiceTarget` doesn't support your desired approach at all.

Can you please provide a reproducer for your problem (via JIRA - https://issues.jboss.org/browse/WFLY-10784)? If we had that, we could start fiddling with it and see what we can do.

Comment (kudrevatykh) [reply weld/subsystem/src/main/java/org/jboss/as/weld/deployment/processors/WeldDeploymentProcessor.java]: I added commit with reproducer, but can't run any test in wildlfy source tree, only in separate project

File: weld/subsystem/src/main/java/org/jboss/as/weld/deployment/processors/WeldDeploymentProcessor.java (hunk 2 above)
Comment (kudrevatykh) [reply weld/subsystem/src/main/java/org/jboss/as/weld/deployment/processors/WeldDeploymentProcessor.java]: fixed

**Comment by manovotn** [comment]
@kabir please close this PR. I will raise a different one with the approach Brian suggested on JIRA.
